sys/drv/rtl8139.c

```diff
@@ -39,28 +48,79 @@ static int rtl_reset(rtl8139_state_t *state) {
   return EAGAIN;
 }
 
+static void do_receive(rtl8139_state_t *state) {
+  /* the device insert their own header which contains size and status of packet
+   */
+  ringbuf_produce(&state->rx_buf, RL_PKTHDR_SIZE);
+  rtl8139_packet_t pkt;
+  ringbuf_getnb(&state->rx_buf, (uint8_t *)&pkt, RL_PKTHDR_SIZE);
+  ringbuf_restorenb(&state->rx_buf, RL_PKTHDR_SIZE);
+  size_t real_len = pkt.len + RL_PKTHDR_SIZE;
+  ringbuf_produce(&state->rx_buf, real_len);
+  /* TODO: SPINLOCK? */
+  if (!ringbuf_movenb(&state->rx_buf, &state->dev_rx_buf, real_len)) {
+    /* discard packet silently */
+    ringbuf_consume(&state->rx_buf, real_len);
+  }
+
+  bus_write_2(state->regs, RL_ISR, RL_ISR_RX_OK);
+}
+
 static intr_filter_t rtl8139_intr(void *data) {
   rtl8139_state_t *state = data;
   uint16_t status = bus_read_2(state->regs, RL_ISR);
 
   if (status & RL_ISR_RX_OK) {
-    bus_write_2(state->regs, RL_ISR, RL_ISR_RX_OK);
-    /* TODO: add enqueue routines */
+    do_receive(state);
     return IF_FILTERED;
   }
 
   return IF_STRAY;
 }
 
+static int rtl8139_read(vnode_t *v, uio_t *uio, int ioflag) {
+  rtl8139_state_t *state = devfs_node_data(v);
+  uio->uio_offset = 0;
+
+  /* TODO: WITH_SPIN/MUTEX_LOCK ? */
+  if (!ringbuf_empty(&state->dev_rx_buf)) {
+    size_t uio_resid = uio->uio_resid;
+    rtl8139_packet_t pkt;
+    ringbuf_getnb(&state->dev_rx_buf, (uint8_t *)&pkt, RL_PKTHDR_SIZE);
```

Comment:
What if inside `ringbuf_getnb` you get interrupt that triggers `do_receive`? Is it safe to do that without any lock?